relay-server/src/utils/garbage.rs

```diff
@@ -0,0 +1,82 @@
+use std::sync::mpsc;
+
+pub struct GarbageDisposal<T> {
+    tx: mpsc::Sender<T>,
+}
+
+impl<T: Send + 'static> GarbageDisposal<T> {
+    pub fn new() -> Self {
+        let (tx, rx) = mpsc::channel();
+
+        std::thread::spawn(move || {
+            relay_log::info!("Start garbage collection thread");
+            while let Ok(object) = rx.recv() {
+                // TODO: Log size of channel queue as a gauge here
+                relay_log::trace!(
+                    "Dropping object {:?} of type {}",
+                    &object as *const T,
+                    std::any::type_name::<T>()
+                );
+                drop(object);
+            }
+
+            // TODO: We create a memory leak if this thread silently crashes
+        });
+
+        Self { tx }
+    }
+
+    pub fn dispose(&self, object: T) {
+        self.tx
+            .send(object)
```

Comment:
Could you run a micro-benchmark to check what the throughput with large project states is? The `Project` struct is _128_ bytes in size, so it could be more efficient to box them since the queue becomes more compact. 

Of course, the allocation of the project still exists then, and we'd have more memory fragmentation; hence a benchmark.

Reply:
I wrote a benchmark here: https://github.com/getsentry/relay/tree/bench/garbage-disposal (did not merge that branch because it uses some quick and dirty imports).

TL;DR boxing the `Project` does not seem to bring an advantage.

```
evict                   time:   [3.6439 us 3.6486 us 3.6536 us]                  
evict_defer             time:   [1.9363 us 1.9732 us 2.0148 us]                         
evict_convert_to_boxes  time:   [2.0021 us 2.0390 us 2.0775 us]                                    
evict_boxes             time:   [2.0992 us 2.2096 us 2.3362 us]
```